accounts/abi/bind/backends: add simulated reorgs

ggwpez opened the pull request:
Adds a Reorg function to the `SimulatedBackend` to enable the simulation of chain reorganizations.  
This brings the possibility to test protocol implementations for the reorg case which only rarely occurs on-chain.

Example:
Assume we have two dependent transactions `A` and `B`. `B` can only be mined after `A` was mined.  
Each transaction is mined in its own block.  
The following code reorders `B` before `A` (with depth=2, length=3), therefore rendering `B` invalid. `A` remains valid.
```go
// Send A and B
// Confirm A (ok)
// Confirm B (ok)
sim.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
    return []types.Transactions{txs[1] /*B*/, txs[0] /*A*/, nil}
})
// Confirm A (ok)
// Confirm B (error)
```

Regarding testing: I added a small test to check that the newly generated chain is correctly positioned on the parent block.  
There seems to be no complete test setup in `simulated_test.go` with event emission tests. I considered adding such as out of scope.  

More extensive testing with event replay and TX reordering was done in an external repo [here](https://github.com/perun-network/go-perun/blob/d69c4fe/backend/ethereum/channel/reorg_test.go#L38) in an ERC20 setup.

MariusVanDerWijden commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -130,6 +134,65 @@ func (b *SimulatedBackend) rollback() {
+		// Add the modified TXs to their new blocks, if any.
+		if i < len(newTXs) {
+			for _, tx := range newTXs[i] {
+				block.AddTx(tx)
```
Please use `block.AddTxWithChain(b.blockchain, tx)` otherwise the BLOCKHASH opcode will panic during execution.

MariusVanDerWijden commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -130,6 +134,65 @@ func (b *SimulatedBackend) rollback() {
+// The slice that is passed to f has length depth, and the one returned
+// must have at least length length.
+// Panics if length is not greater than depth.
+func (b *SimulatedBackend) Reorg(ctx context.Context, depth, length int64, f Reorderer) error {
```
F could be inlined (doesn't have to be, but would fit better with the rest of geth).
Also please give f a more descriptive name (`reorder` maybe)

MariusVanDerWijden commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -130,6 +134,65 @@ func (b *SimulatedBackend) rollback() {
+		panic("depth must be greater than 0")
+	}
+	// The chain inserter always chooses the most difficult chain, so a
+	// shorter chain would not trigger a reorg.
```
That's not entirely true, a shorter chain can trigger a reorg if it's difficulty is higher, but I think we can ignore that for the simulated backend. We can change that if someone complains later.

ggwpez replied:
Yes right, i tried to change the difficulty of the inserted blocks by using `block.SetDifficulty` but then they fail to verify with:  
`inserting new chain: invalid difficulty: have 111111111111, want 131072`.  
Maybe the mocked hasher or the difficulty consensus is not happy then.  
  
I will change the comment to:  
```pre
// The chain inserter always chooses the most difficult chain.
// In the simulated case we expect that a longer chain is always
// more difficult and will therefore be picked in a reorg case.
// This check makes the function deterministic in the length<=depth case.
```

MariusVanDerWijden commented on accounts/abi/bind/backends/simulated_test.go:
```diff
@@ -1114,3 +1114,41 @@ func TestSimulatedBackend_CallContractRevert(t *testing.T) {
 	}
 }
+
+func TestSimulatedBackend_Reorg(t *testing.T) {
```
Can you add a short comment about the intention of the test?
Something like
```
TestSimulatedBackend_Reorg tests the reorg functionality.
Hereby 5 blocks are inserted into the chain and a reorg of 
the last 4 blocks to 6 different blocks occurs.
Resulting in a chain of 7 blocks.
```

ggwpez replied:
I describe the steps now:
```pre
// TestSimulatedBackend_ReorgTx tests the Reorg capability of the simulated
// backend. It runs these steps:
// 1. Setup `n` accounts and fund them. This is necessary since the account
//    nonce prevents transactions of the same account to from being re-ordered.
// 2. Deploy the `Counter` contract.
// 3. Send `n` transactions and confirm them.
// 4. Use a permutation to re-order the transactions in a simulated chain Reorg.
// 5. Emulate the `counter` variable of the contract and check that the
//    transactions are correctly confirmed or reverted.
```

MariusVanDerWijden commented on accounts/abi/bind/backends/simulated_test.go:
```diff
@@ -1114,3 +1114,41 @@ func TestSimulatedBackend_CallContractRevert(t *testing.T) {
+	// Assert chain length.
+	if sim.blockchain.CurrentBlock().NumberU64() != 7 {
+		t.Error("wrong length of new chain")
+	}
```
Could you either add a second test, or extend this one to actually shuffle some transactions? 
Would be nice to see that this functionality works not just with empty blocks

ggwpez replied:
Yes currently it is a bit boring :laughing:. I will try to add a contract test.

MariusVanDerWijden reviewed (changes requested):
Hey @ggwpez! Thanks for your contribution. The idea generally looks good to me, added some comments (mostly nitpicks).

ggwpez commented:
So i now:  
- Added a proper reorg test.
- Added a comment that mentions that transactions from the same account can not be reordered because of their nonce.  

Will assign it back to @MariusVanDerWijden for review.

fjl commented:
We looked at this during our PR triage today. @karalabe says it would be simpler to add a `SetHead` function to the backend instead. That way, one could simply reset the test chain to an earlier block, then add the transactions in different order.

ggwpez commented:
> We looked at this during our PR triage today. @karalabe says it would be simpler to add a `SetHead` function to the backend instead. That way, one could simply reset the test chain to an earlier block, then add the transactions in different order.

As far as i know (please correct me);  
This would not trigger the reorg logic in `core/BlockChain.reorg(oldBlock, newBlock *types.Block)`, which is needed to emit the *deleted* and *rebirth* events.  
Having these events is necessary when testing reorg-resistant event subscriptions.  
@fjl @karalabe

matthiasgeihs commented:
Any news on this one? @fjl

holiman commented:
We have some suggestions, @karalabe will write a more thorough description of what can be done here

karalabe commented:
So, we've been thinking about it a bit. You are right that in order to test reorgs, you can't just "setHead", since that would be a blind rollback, not a true reorg.

That said, we still kind of oppose your solution because it seems like a very complex single-shot method that is very very hard to compose. It also doesn't really fit into the call model of the simulated backend where you simply have blocking method calls that sequentially build up the chain.

Our counter-proposal would be to provide a new method analogous to `rollback`, just one that takes a past block (e.g. via hash), which would start constructing a new block on top of that. That would allow the exact current workflow to be used for constructing the reorg too, but once your new chain exceeds the old one, you'd get all the executions and reorged events.

Something akin to:

```go
// Create the original chain
sim.SendTransaction(tx1)
sim.Commit()
sim.SendTransaction(tx2)
sim.Commit()
sim.SendTransaction(tx3)
sim.Commit()

// Start a sidechain to reorg to
sim.Fork(sim.HeaderByNumber(1).Hash) // +error handling
sim.SendTransaction(tx3)
sim.Commit()
sim.Commit()
sim.Commit() // At this point your fork becomes heavier and should trigger the reorg even you want
```

Would this work? Seems like a much cleaner approach. You might need to massage `Commit/rollback` too a bit to take into consideration the last committed block instead of defaulting to the `CurrentBlock`.

ggwpez commented:
Thanks for the long reply!  
I will prototype your suggestion and then report back here how/if it works.

holiman commented:
> The restriction is to use only use SendTransaction and Commit after a fork, since all other function (like CallContract or Rollback) still use CurrentBlock and operate on the canonical chain.

On a 'normal' node, a sidechain import does not actually execute the sideblocks until/unless they overtake the canon chain. Thus, it's not actually possible to do CallContract on a side-state, so that's definitely ok. Same with rollback, you can't really roll back a sidechain. Theoretically, if you roll back to number X, then it _might_ be possible for the TD of `X` to be higher on the sidechain than it was on the canon chain, not sure what would happen in that case.

Anyway, both things seem ok to me

ggwpez commented:
Ok great. Who can review this?

karalabe reviewed (approved):
Made some minor tweaks. LGTM now

MariusVanDerWijden reviewed (approved):
LGTM
